pkg/template/api/helpers.go

```diff
@@ -17,21 +17,27 @@ func AddObjectsToTemplate(template *Template, objects []runtime.Object, targetVe
 			return errors.New("cannot add a nil object to a template")
 		}
 
-		kind, _, err := kapi.Scheme.ObjectKind(obj)
+		// We currently add legacy types first to the scheme, followed by the types in the new api
+		// groups. We have to check all ObjectKinds and not just use the first one returned by
+		// ObjectKind().
+		gvks, _, err := kapi.Scheme.ObjectKinds(obj)
 		if err != nil {
 			return err
 		}
 
 		var targetVersion *unversioned.GroupVersion
```

Comment:
This logic is really confused at this point - Codecs.CodecForVersions accepts a runtime.GroupVersioner (which gvks satisfies) which should be enough to ensure you get the right object, so you shouldn't have to pass a single target version (you can instead pass multiple).  And it raises the question about whether we should even be using this logic for templates vs dealing with unstructured.

However, I guess it can stand for now.